mailserver/mailserver.go

```diff
@@ -57,43 +67,59 @@ func NewDbKey(t uint32, h common.Hash) *DBKey {
 }
 
 // Init initializes mailServer
-func (s *WMailServer) Init(shh *whisper.Whisper, path string, password string, pow float64) error {
+func (s *WMailServer) Init(shh *whisper.Whisper, config *params.WhisperConfig) error {
 	var err error
+	path := filepath.Join(config.DataDir, "mailserver", "data")
 	if len(path) == 0 {
-		return fmt.Errorf("DB file is not specified")
+		return fmt.Errorf("data directory not provided")
 	}
 
-	if len(password) == 0 {
+	if len(config.Password) == 0 {
 		return fmt.Errorf("password is not specified")
 	}
 
 	s.db, err = leveldb.OpenFile(path, nil)
 	if err != nil {
-		return fmt.Errorf("open DB file: %s", err)
+		return fmt.Errorf("open DB: %s", err)
 	}
 
 	s.w = shh
-	s.pow = pow
+	s.pow = config.MinimumPoW
 
-	MailServerKeyID, err := s.w.AddSymKeyFromPassword(password)
+	MailServerKeyID, err := s.w.AddSymKeyFromPassword(config.Password)
 	if err != nil {
 		return fmt.Errorf("create symmetric key: %s", err)
 	}
 	s.key, err = s.w.GetSymKey(MailServerKeyID)
 	if err != nil {
 		return fmt.Errorf("save symmetric key: %s", err)
 	}
+	s.rateLimit = config.RateLimit
+	s.setupMailServerCleanup(config.MailServerCleanupPeriod)
+
 	return nil
 }
 
+func (s *WMailServer) setupMailServerCleanup(period int) {
+	if s.rateLimit <= 0 || period <= 0 {
+		return
+	}
+	if s.tick == nil {
+		s.tick = &ticker{}
+	}
+	d := time.Duration(period) * time.Second
+	go s.tick.run(d, s.removeExpiredRateLimits)
```

Comment:
It can be done without another struct:

```go
go s.startRateLimitsCleanup(duration)

// 
func (s *WMailServer) startRateLimitsCleanup(tick time.Duration) {
  s.wg.Add(1)
  defer s.wg.Done()

  for {
    select {
    case <-s.quit:
    case <-time.After(tick):
      s.removeExpiredRateLimits()
    }
  }
}
```

Reply:
yep, initially used a different struct to also support old message removal. I think @pilu is still thinking on using it, otherwise i'm fine to switch to this approach.